### PR TITLE
Included 4d and 5d tests for elementwise ops

### PR DESCRIPTION
**Pull request by meena-ibm**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [ x ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Included the 4d and 5d tests for the below element-wise operators in test_inductor_ops.py

Torch.clamp
torch.add_
torch.mul_
torch.copy_
 torch.cos
 torch.sin
torch.ne
 torch.le
torch.eq
 torch.neg
torch.nn.functional.silu
 torch.sigmoid
 torch.nn.gelu
 torch.nn.functional.softplus
torch.add
torch.sub
 torch.mul
 torch.div,
torch.true_divide
 
Test results:
=============================== test session starts ===============================
platform linux -- Python 3.12.12, pytest-9.0.3, pluggy-1.6.0 -- /opt/dti-venv/bin/python3
cachedir: .pytest_cache
rootdir: /torch-spyre-dev
configfile: pyproject.toml
collected 693 items / 475 deselected / 218 selected

================= 218 passed, 475 deselected in 328.30s (0:05:28) =================

### CODE REVIEW

**Comment by meena-ibm** [comment]
@dgrove-oss This is same changes related to PR https://github.com/torch-spyre/torch-spyre/pull/1887. Since unexpected commits went in the previous PR, created this new clean PR.

**Comment by dgrove-oss** [comment]
Thanks for creating the fresh PR.

The new test `test_pointwise_range_op_clamp_5d_empty_spyre` is failing in CI.

Please fix the linter failures.

One small suggestion from code review:

SUGGESTION: Several added entries use double-wrapped shape tuples (e.g., cached_randn(((5, 10, 64, 128, 128)), ...)) — the extra parens are harmless but inconsistent with the rest of the file.

**Comment by meena-ibm** [comment]
@dgrove-oss I have removed the failing test for now, will do the analysis and raise issue if required. Sorry I didn't notice the double-wrapped shape tuples, corrected it with recent commit. Thanks!

**Comment by dgrove-oss** [comment]
@meena-ibm  -- I think the last commit to resolve the linter failure didn't have a proper verified signature.  Please fix that when you can and we should be able to merge this.

**Comment by meena-ibm** [comment]
> @meena-ibm -- I think the last commit to resolve the linter failure didn't have a proper verified signature. Please fix that when you can and we should be able to merge this.

done,please check

**Review by dgrove-oss — DISMISSED** [review]
lgtm

**Comment by dgrove-oss** [comment]
We have repeatable CI failures now that need to be resolved before this can be merged.

**Comment by dgrove-oss** [comment]
> We have repeatable CI failures now that need to be resolved before this can be merged.

@meena-ibm -- do you want to resolve the merge conflict and pursue the CI failures or are we going to abandon the PR?

**Comment by meena-ibm** [comment]
@dgrove-oss sorry I somehow missed your previous message, I will resolve the merge conflict and check the CI failures.

**Comment by meena-ibm** [comment]
@dgrove-oss Merge conflicts has been resolved, recent push dismissed your review, please review it. I am unable to trigger the tests. Please trigger it,  I will check the test status once the run is completed.

File: tests/inductor/test_inductor_ops.py [hunk 1]
```diff
@@ -2064,6 +2130,42 @@ def setUp(self):
...
+                    torch.zeros((2, 3, 16, 112, 112), dtype=torch.float16),
+                ),
+                "5d_inf": (
+                    torch.full((2, 3, 4, 5, 6), float("inf"), dtype=torch.float16),
+                    torch.ceil(cached_randn((2, 3, 4, 5, 6), abs=True, scale=9.9)).to(
+                        dtype=torch.float16
+                    ),
+                ),
+                "5d_empty": (
+                    torch.empty((2, 3, 4, 5, 6), dtype=torch.float16),
```
Comment (dgrove-oss) [inline tests/inductor/test_inductor_ops.py]: This test is faulty.  By definition `torch.empty` returns uninitiated storage.  Using it without first initializing it is undefined.

File: tests/inductor/test_inductor_ops.py [hunk 2]
```diff
@@ -871,6 +887,55 @@ def setUp(self):
...
+                "5d_neg": (-torch.randn((1, 3, 1, 64, 64), dtype=torch.float32),),
+                "5d_nan": (
+                    torch.full(
+                        (16, 64, 64, 128, 128), float("nan"), dtype=torch.float16
+                    ),
+                ),
+                "5d_inf": (
+                    torch.full((1, 1, 4, 4, 4), float("inf"), dtype=torch.float32),
+                ),
+                "5d_empty": (torch.empty((1, 1, 1, 1, 1), dtype=torch.float16),),
```
Comment (dgrove-oss) [inline tests/inductor/test_inductor_ops.py]: This test is faulty.  By definition `torch.empty` returns uninitiated storage.  Using it without first initializing it is undefined.

**Review by dgrove-oss — CHANGES_REQUESTED** [review]
Summary

Adds 4D/5D parameterized test coverage in tests/inductor/test_inductor_ops.py for a batch of elementwise ops (clamp, add_/mul_/copy_, cos/sin, ne/le/eq/gt/ge, neg, silu, sigmoid, gelu, softplus, add/sub/mul/div/true_divide). Single file changed, +313/-1.

Review

BLOCKERs

- [ ] Several new param sets feed torch.empty(...) (uninitialized memory) directly into the op under test, and this already fails deterministically in the PR's own CI. gh pr checks 2644 shows test_cmp_ge_5d_empty_spyre and test_cmp_gt_5d_empty_spyre failing with Mismatched elements: 22 / 720 (3.1%), reproduced identically across the initial run and the pod-level retry, in both Test Inductor Ops Pointwise and Test Inductor Ops Lx Pointwise. These map to the "5d_empty" case added to ("test_cmp", "test_binary_op_cpu") (diff lines ~2163-2167), where x = torch.empty((2, 3, 4, 5, 6), dtype=torch.float16) is the left operand of torch.ge/torch.gt against a torch.ceil(cached_randn(...)) right operand.

  Regardless of the exact mechanism, comparing against — or otherwise computing with — the contents of uninitialized memory is not a valid test case: torch.empty's contents are unspecified by PyTorch's own docs, so a test built on them isn't asserting anything about the op's correctness, and any expectation of a specific output value (including "should match CPU") isn't well-founded to begin with, independent of whether it happens to pass or fail in a given environment. That it already fails in CI just makes the point concrete.

  Every read of torch.empty(...)'s tensor contents introduced by this PR must be removed, not just the ones already observed failing:
  - "5d_empty" in test_cos_sin/test_unary_op (diff line 85) — cos/sin applied to torch.empty((1,1,1,1,1)).
  - "5d_empty" in test_cmp/test_binary_op_cpu (diff line 134) — the one currently failing in CI.
  - "5d_with_ones" and "5d_with_empty" in test_inplace_op/test_inplace_op_cpu (diff lines 317, 319-330) — torch.empty used as the second operand (src) of add_/mul_.
  - "5d_empty" in the op block around diff line 353-356 — both operands are torch.empty(...).

  Each should be replaced with a tensor whose values are actually specified (e.g. torch.zeros, torch.ones, cached_randn, or a torch.full sentinel), so the test both passes for a defensible reason and documents what it's actually exercising. If the intent was to test the op's handling of unspecified/don't-care memory (a legitimate but different goal, e.g. verifying an in-place write fully overwrites a buffer rather than reading it), that needs a different test shape entirely (e.g. asserting only on the parts of the output that are actually defined), not a compare_with_cpu value comparison.
- [ ] test_softplus_5d_nan_* also fails deterministically in CI (Test Inductor Ops Lx Misc Compute B, 100% mismatch — 720/720 and 360/360 elements). This isn't a torch.empty case (it's torch.full(..., float('nan'))), so it's a separate defect from the one above, but it's still a new test case from this PR failing in the PR's own CI and must be fixed or the case removed/marked as a known-issue placeholder (the PR already uses that pattern elsewhere, e.g. the commented-out # To be enabled when #1833 is resolved cases) before merge.
- [ ] The PR description's reported test results ("218 passed... 0:05:28") do not reflect the current state of the branch — CI now shows multiple failures. The description should be updated, or the discrepancy explained, before merge; as-is it's misleading about the PR's actual status.

SUGGESTIONs

- [ ] Several commented-out test cases reference open issues (#707, #1851, #1833, #1774, #1831, #1829, #1830, #442, #1882) as reasons they're disabled. Worth double-checking each is still open/relevant rather than stale, but not blocking.
- [ ] test_softplus_cpu(self, x, beta) → test_softplus_cpu(self, x, beta=1.0) (diff line ~406) — giving beta a default changes the method's calling contract; confirm no other param sets relied on beta being required, and that this wasn to allow the new no-beta cases added in this PR call the method with fewer args intentionally rather than by accident.
- [ ] Given the volume of # To be enabled when #NNNN is resolved commented-out blocks this PR accumulates, consider whether a pytest.mark.xfail(reason=...) per case (rather than commenting the code out) would keep the cases visible and automatically re-enable/flag when the underlying bug is fixed, instead of relying on someone remembering to uncomment them later.

QUESTIONs

- [ ] None beyond the above — the failures are concrete and don't need author clarification to act on.

Verdict

REQUEST_CHANGES — two BLOCKERs on test validity/correctness (uninitialized-memory reads in test inputs, one of which already manifests as a deterministic CI failure; a second, unrelated deterministic CI failure in the NaN softplus case), plus a description that no longer matches CI reality.

**Comment by meena-ibm** [comment]
@dgrove-oss Below changes are made as per the review comments:

- removed tests using torch.empty()
- changes done in utils_inductor.py and contest.py to accommodate expect_fail with reason
- marked all commented tests to xfail with issue number
- beta changes made in test_softmax is intended and it doesn't affect other tests. confirmed by running all relevant tests.
- updated the description of the PR with the recent test results
- test_softplus_5d_nan- this test is passing so didn't make any changes here

Please let me know if any other changes are required. Thank you!
